app/models/purchase.rb

```diff
@@ -239,6 +239,13 @@ def self.affair_class
     "#{name}Affair".constantize
   end
 
+  def bookkeepable?
+    with_accounting &&
+    invoice? &&
+    items.any? &&
+    TaxDeclarationItemPart.where(journal_entry_item_id: journal_entries.last.items.map(&:id)).any?
```

Comment:
Prefer `.select` instead of `.map` here because Arel will build only one query, that's the best!